File: BankingAppObjectiveC/BankingApp/OnlineMeetingViewController.m
//+----------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// Module name: OnlineMeetingViewController.m
//+----------------------------------------------------------------


#import "OnlineMeetingViewController.h"
#import "ChatViewController.h"
#import "VideoViewController.h"
#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>



@interface OnlineMeetingViewController ()
@property (weak, nonatomic) IBOutlet UITextView *meetingURL;
@property (weak, nonatomic) IBOutlet UITextField *displayName;
@property (weak, nonatomic) IBOutlet UIButton *join;
@property (weak, nonatomic) IBOutlet UILabel *tokenAndDiscoveryURISuccessLabel;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicatorForServiceApplicationResponse;

typedef void (^completionBlock)( NSData *data ,NSError *error);
@end

@implementation OnlineMeetingViewController


SfBConversation *onlineMeetingConversation ;
NSString *token;
NSString *discoveryURI;

#pragma mark - Lifecycle and helper functions
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    // Configure Shared application instance for Online meeting
        
    // Setup UI
    self.join.titleLabel.textAlignment = NSTextAlignmentCenter;
    self.displayName.text = [Util getMeetingDisplayName];
    self.displayName.delegate = self;
    }

- (void)viewDidAppear:(BOOL)animated{
     // First POST request to fetch meeting URL
    [self sendPostRequestForMeetingURL];
    self.navigationController.navigationBarHidden = false;
}

#pragma mark - Send POST requests for online meeting call flow functions
//  POST request to fetch ad hoc Meeting URL
-(void) sendPostRequestForMeetingURL{
// request to Trusted Application API Service Application endpoint
    NSString* meetingUrlRequestString = [Util getOnlineMeetingRequestURL];
    NSMutableURLRequest *meetingUrlRequest = [[NSMutableURLRequest alloc] initWithURL:[[NSURL alloc] initWithString:meetingUrlRequestString]];

    
    
    meetingUrlRequest.HTTPMethod = @"POST";
    meetingUrlRequest.HTTPBody = [@"Subject=adhocMeeting&Description=adhocMeeting&AccessLevel=" dataUsingEncoding:NSUTF8StringEncoding];
    [self SendHttpRequest:meetingUrlRequest withBlock:^(NSData *data, NSError *error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if(error == nil){
                
                NSError *e = nil;
                NSDictionary *json = [NSJSONSerialization JSONObjectWithData: data options: NSJSONReadingMutableContainers error: &e];
                
                if (!json) {
                    NSLog(@"Error parsing JSON: %@", e);
                } else {
                    self.meetingURL.text = json[@"JoinUrl"];
                     NSLog(@"Successful! meeting URL>>%@", json[@"JoinUrl"]);
                    //Send second POST request to get discovery URI and TOKEN based on response meeting URL
                    [self sendPostRequestForTokenAndDiscoveryURI];

                }
            }
            else{
                NSLog(@"ERROR! Getting meeting URL failed>%@",error);
                
                [Util showErrorAlert:error inView:self];
                
            }
        });
    }];

}

//  POST request to get token and discovery URI based on response meeting URL
-(void) sendPostRequestForTokenAndDiscoveryURI  {
     // request to Trusted Application API Service Application endpoint
    
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc]initWithURL:([[NSURL alloc]initWithString:[Util getTokenAndDiscoveryURIRequestURL]])];
    request.HTTPMethod = @"POST";
    NSString * HTTPBodyString = [NSString stringWithFormat:@"ApplicationSessionId=AnonMeeting&AllowedOrigins=http%%3a%%2f%%2flocalhost%%2f&MeetingUrl=%@", self.meetingURL.text];
    request.HTTPBody = [HTTPBodyString dataUsingEncoding:NSUTF8StringEncoding];
    self.join.enabled = false;
    [self SendHttpRequest:request withBlock:^(NSData *data, NSError *error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if(error == nil){
                
                NSError *e = nil;
                NSDictionary *json = [NSJSONSerialization JSONObjectWithData: data options: NSJSONReadingMutableContainers error: &e];
                
                if (!json) {
                    NSLog(@"Error parsing JSON: %@", e);
                } else {
                    
                    discoveryURI = json[@"DiscoverUri"];
                    self.tokenAndDiscoveryURISuccessLabel.textColor = [UIColor darkGrayColor];
                    self.tokenAndDiscoveryURISuccessLabel.text = @"Success! Please join online meeting";
                    token = json[@"Token"];
                    self.join.enabled = true;
                    self.join.alpha = 1;
                    [self.activityIndicatorForServiceApplicationResponse stopAnimating];
                    NSLog(@"Successful! token and discovery URI:%@, %@",json[@"Token"],json[@"DiscoverUri"]);
                }
            }
            else{
                NSLog(@"ERROR! Getting token and discovery URI failed>%@",error);
               [Util showErrorAlert:error inView:self];
                 
            }
        });
    }];

    
}

// Helper function to send request
-(void)SendHttpRequest: (NSURLRequest *)request withBlock: (completionBlock) completionHandler{
    NSURLSession *sessionObject = [NSURLSession sharedSession];
    NSURLSessionDataTask *task = [sessionObject dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSInteger statusCode = 0;
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        statusCode = httpResponse.statusCode;
       NSLog(@"statusCode-> %ld", (long)statusCode);
        if(statusCode == 200){
            completionHandler(data,nil);
        }
        else{
            if(error == nil){
                NSError* myError = [[NSError alloc]initWithDomain:@"custom error" code:statusCode userInfo:nil ];
                
               
                completionHandler( nil, myError);
            }
            else {
                completionHandler( nil, error );
            }
        }
        
    }];
    [task resume];
}

#pragma mark - Join online meeting anonymous with discover URI and token function

-(bool) didJoinMeeting  {
    NSError *e = nil;
     SfBConversation *conversationInstance =   [ _onlineMeetingsfb joinMeetingAnonymousWithDiscoverUrl:[[NSURL alloc] initWithString:discoveryURI] authToken:token displayName:self.displayName.text error:&e].conversation;
    if(conversationInstance){
    onlineMeetingConversation = conversationInstance;
    
        return true;
    }
    else{
        NSLog(@"ERROR! Joining online meeting>%@",e);
        
              [Util showErrorAlert:e inView:self];
        }
    return false;
}

- (void)didReceiveAlert:(SfBAlert *)alert{
    [alert showSfBAlertInController:self];
}
#pragma mark -  User button actions
// press "Join online meeting" button to join text or video online meeting
- (IBAction)JoinOnlineMeetingPressed:(UIButton *)sender {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Join online meeting" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Join online chat" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self joinOnlineChat];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Join online video chat" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self joinOnlineVideoChat];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    UIPopoverPresentationController *popoverController =  alertController.popoverPresentationController;
    if(popoverController != nil){
        popoverController.sourceView = sender;
        popoverController.sourceRect = sender.bounds;
        
    }
    
    [self presentViewController:alertController animated:YES completion:nil];
}

-(void)joinOnlineChat{
    if([self didJoinMeeting]){
        [self performSegueWithIdentifier:@"joinOnlineChat" sender:nil];
        
    }
}

-(void)joinOnlineVideoChat{
    
    SfBConfigurationManager *config = _onlineMeetingsfb.configurationManager;
    NSString *key = @"AcceptedVideoLicense";
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    
    if([defaults boolForKey:key]){
        [config setEndUserAcceptedVideoLicense];
        if([self didJoinMeeting]){
            [self performSegueWithIdentifier:@"joinOnlineAudioVideoChat" sender:nil];
        }
    }else{
        MicrosoftLicenseViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:@"MicrosoftLicenseViewController"];
        vc.delegate = self;
        [self presentViewController:vc animated:YES completion:nil];
    }
}

 #pragma mark -Segue navigation functions
-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender{
    if([segue.identifier isEqualToString:@"joinOnlineChat"]){
        ChatViewController* destination = segue.destinationViewController;
        destination.conversation = onlineMeetingConversation;
        
    }
    else if([segue.identifier isEqualToString:@"joinOnlineAudioVideoChat"]){
        VideoViewController* destination = segue.destinationViewController;
        destination.deviceManagerInstance = _onlineMeetingsfb.devicesManager;
        destination.conversationInstance = onlineMeetingConversation;
        destination.displayName = self.displayName.text;
        
        discoveryURI = nil;
        token = nil;
    }
     onlineMeetingConversation = nil;
}


#pragma mark -  Lifecycle and helper functions
// reset UI when leaving this screen

-(void)viewDidDisappear:(BOOL)animated{
    self.displayName.text = self.displayName.placeholder;
    self.meetingURL.textColor = [UIColor redColor ];
    self.meetingURL.text = @"Waiting for online meeting URL!";
    self.tokenAndDiscoveryURISuccessLabel.textColor = [UIColor redColor ];
    self.tokenAndDiscoveryURISuccessLabel.text = @"Waiting for token and discovery URI!";
    
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField{
    [self.view endEditing:true];
    return false;
}

#pragma mark - MicrosoftLicenseViewController delegate function
- (void)controller:(MicrosoftLicenseViewController* )controller
  didAcceptLicense:(BOOL)acceptedLicense{
    if(acceptedLicense){
        SfBConfigurationManager *config = _onlineMeetingsfb.configurationManager;
        [config setEndUserAcceptedVideoLicense];
        if([self didJoinMeeting]){
            [self performSegueWithIdentifier:@"joinOnlineAudioVideoChat" sender:nil];
        }
    }
}


@end

File: BankingAppSwift/BankingAppSwift/OnlineMeetingViewController.swift
//+----------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// Module name: OnlineMeetingViewController.swift
//----------------------------------------------------------------
/**
 *  OnlineMeetingViewController implements the online meeting call flow using client-side
 *  functionality of Skype for Business App SDK.
 *  You will require Trusted Application API Service Application(SaaS app) for this work-flow.
 *
 */



import Foundation
import UIKit
import SkypeForBusiness


class OnlineMeetingViewController: UIViewController, SfBAlertDelegate,UITextFieldDelegate, MicrosoftLicenseViewControllerDelegate {
    /** Called when new alert appears in the context where this delegate is attached.
     *
     * Each alert is passed to a delegate once and dismissed unconditionally.
     * If no delegate is attached, alerts are accumulated and reported as soon
     * as delegate is set. Accumulated alerts of the same category and type
     * are coalesced, only the last one will be reported.
     */

    
    var sfb: SfBApplication?
    fileprivate var conversation: SfBConversation?
    fileprivate var token: String?
    fileprivate var discoveryURI: String?
    
    @IBOutlet var meetingUrl: UITextView!
    @IBOutlet var displayName: UITextField!
    @IBOutlet var join: UIButton!
    @IBOutlet var tokenAndDiscoveryURISuccessLabel: UILabel!
    
    @IBOutlet weak var activityIndicatorForServiceApplicationResponse: UIActivityIndicatorView!
    //MARK: Lifecycle and helper functions
    override func viewDidLoad() {
        super.viewDidLoad()
        // Setup UI
        join.titleLabel?.textAlignment = NSTextAlignment.center
        displayName.text = getMeetingDisplayName
        displayName.delegate = self
    }
    
    override func viewWillAppear(_ animated: Bool) {
        
        self.navigationController?.setNavigationBarHidden(false, animated: false)
        
    }
    
    override func viewDidAppear(_ animated: Bool) {
        // First POST request to fetch meeting URL
        sendPostRequestForMeetingURL()
        
    }
    
    //MARK:Send POST requests for online meeting call flow functions
    //  POST request to fetch ad hoc Meeting URL
    func sendPostRequestForMeetingURL(){
        
        // request to Trusted Application API Service Application endpoint
        if let url = URL(string: getOnlineMeetingRequestURL){
        let meetingUrlRequest = NSMutableURLRequest(url: url)
        meetingUrlRequest.httpMethod = "POST"
        meetingUrlRequest.httpBody = "Subject=adhocMeeting&Description=adhocMeeting&AccessLevel=".data(using: String.Encoding.utf8)
        
        SendHttpRequest(meetingUrlRequest as URLRequest) { (data, error) in
            DispatchQueue.main.async {
                do {
                    guard error == nil, let data = data else {
                        throw error!
                    }
                    
                    let json = try  JSONSerialization.jsonObject(with: data, options: []) as! [String: String]
                    self.meetingUrl.text = (json["JoinUrl"])!;
                    
                    print("Successful! meeting URL>> \(json["JoinUrl"])");
                    
                    // Send second POST request to get discovery URI and TOKEN based on response meeting URL
                    self.sendPostRequestForTokenAndDiscoveryURI();
                } catch {
                    print("ERROR! Getting meeting URL failed>\(error)")
                    showErrorAlert("Getting meeting URL failed. Try again later!", viewController: self)
                   
                }
                
                
            }
        }
        }
    }
    
    //  POST request to get token and discovery URI based on response meeting URL
    func sendPostRequestForTokenAndDiscoveryURI()  {
        
        // request to Trusted Application API Service Application endpoint
         if let url = URL(string: getTokenAndDiscoveryURIRequestURL){
        let request = NSMutableURLRequest(url: url)
        request.httpMethod = "POST"
        let meetintURL = meetingUrl.text!;
        request.httpBody = "ApplicationSessionId=AnonMeeting&AllowedOrigins=http%3a%2f%2flocalhost%2f&MeetingUrl=\(meetintURL)".data(using: String.Encoding.utf8)
        
        join.isEnabled = false
        SendHttpRequest(request as URLRequest) { (data, error) in
            DispatchQueue.main.async {
                do {
                    guard error == nil, let data = data else {
                        throw error!
                    }
                    
                    let json = try  JSONSerialization.jsonObject(with: data, options: []) as! [String: String]
                    self.discoveryURI = json["DiscoverUri"]!
                    self.tokenAndDiscoveryURISuccessLabel.textColor = UIColor.darkGray
                    self.tokenAndDiscoveryURISuccessLabel.text = "Success! Please join online meeting"
                    self.token = json["Token"]
                    self.join.isEnabled = true
                    self.join.alpha = 1
                    self.activityIndicatorForServiceApplicationResponse.stopAnimating()
                    print("Successful! token and discovery URI>> \(json["Token"]),\(json["DiscoverUri"])");
                    
                } catch {
                    print("ERROR! Getting token and discovery URI failed>\(error)")
                    showErrorAlert("Getting Discover URI failed. Try again later!", viewController: self)
                    
                    
                }
                
                
            }
        }
     }
        
    }
    
    
    // Helper function to send request
    func SendHttpRequest(_ request:URLRequest, withBlock completionHandler:  @escaping completionBlock) {
        let  sessionObject: URLSession = URLSession.shared
        let task = sessionObject.dataTask(with: request, completionHandler: {  (data, response, error) in
            let httpResponse = response as? HTTPURLResponse
            let statusCode = httpResponse?.statusCode
            print("statusCode-aas>\(statusCode)")
            
            if (statusCode == 200 ) {
                
                completionHandler(data, nil)
                
            }
                
            else{
                if(error == nil){
                    let error:NSError = NSError(domain: "statusCode -\(statusCode)!", code: statusCode!, userInfo:nil)
                    completionHandler(nil,error as NSError?)
                }
                else {
                    completionHandler(nil,error as NSError?)
                }
                
            }
        }) 
        
        task.resume()
    }
    
    //MARK: User button actions
    // press "Join online meeting" button to join text or video online meeting
    @IBAction func JoinOnlineMeetingPressed(_ sender: AnyObject) {
        let alertController:UIAlertController = UIAlertController(title: "Join online meeting", message: nil, preferredStyle: .actionSheet)
        
        
        alertController.addAction(UIAlertAction(title: "Join online chat", style: .default, handler: { (action:UIAlertAction) in
            self.joinOnlineChat()
        }))
        
        alertController.addAction(UIAlertAction(title: "Join online video chat", style: .default, handler: { (action:UIAlertAction) in
            self.joinOnlineVideoChat()
        }))
        
        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        
        
        if let popoverController = alertController.popoverPresentationController {
            popoverController.sourceView = sender as? UIView
            popoverController.sourceRect = sender.bounds
        }
        self.present(alertController, animated: true, completion: nil)
    }
    
    func joinOnlineChat()  {
        if(didJoinMeeting()){
            self.performSegue(withIdentifier: "joinOnlineChat", sender: nil)
        }
    }
    
    func joinOnlineVideoChat()  {
        
        
        if let sfb = sfb{
            let config = sfb.configurationManager
            let key = "AcceptedVideoLicense"
            let defaults = UserDefaults.standard
            
            if defaults.bool(forKey: key) {
                config.setEndUserAcceptedVideoLicense()
                if(didJoinMeeting()){
                    self.performSegue(withIdentifier: "joinOnlineAudioVideoChat", sender: nil)
                }
                
            } else {
                
                
                let vc = self.storyboard?.instantiateViewController(withIdentifier: "MicrosoftLicenseViewController") as! MicrosoftLicenseViewController
                vc.delegate = self
                
                self.present(vc, animated: true, completion: nil)
            }
            
        }
        
        
        
    }
    
    //MARK: Join online meeting anonymous with discover URI and token function
    
    func didJoinMeeting() -> Bool {
        do {
            
            let session = try sfb!.joinMeetingAnonymous(withDiscover: URL(string: discoveryURI!)!, authToken: token!, displayName: displayName.text!)
            conversation = session.conversation
            conversation?.alertDelegate = self
            return true
        } catch {
            print("ERROR! Joining online meeting>\(error)")
            showErrorAlert("Joining online meeting failed. Try again later!", viewController: self)
            return false
        }
    }
    
    //MARK: Segue navigation functions
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if(segue.identifier == "joinOnlineChat"){
            guard let destination = segue.destination as? ChatViewController else {
                return
            }
            destination.conversation = self.conversation
            
        }
        else if(segue.identifier == "joinOnlineAudioVideoChat"){
            guard let destination = segue.destination as? VideoViewController else {
                return
            }
            destination.deviceManagerInstance = sfb!.devicesManager
            destination.conversationInstance = conversation
            destination.displayName = displayName.text!
            
            discoveryURI = nil
            token = nil
            
        }
        conversation = nil
    }
    
    //MARK: Lifecycle and helper functions
    // reset UI when leaving this screen
    override func viewDidDisappear(_ animated: Bool) {
        displayName.text = displayName.placeholder
        meetingUrl.textColor = UIColor.red
        meetingUrl.text = "Waiting for online meeting URL!"
        tokenAndDiscoveryURISuccessLabel.textColor = UIColor.red
        tokenAndDiscoveryURISuccessLabel.text = "Waiting for token and discovery URI!"
    }
    
    // Hide Keyboard on return.
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        self.view.endEditing(true)
        return false
    }
    
    
    //MARK: SfBAlertDelegate alert function
    func didReceive(_ alert: SfBAlert) {
       alert.showSfBAlertInController(self)
    }
    
    func controller(_ controller: MicrosoftLicenseViewController, didAcceptLicense acceptedLicense: Bool) {
        if(acceptedLicense){
            if let sfb = self.sfb{
                let config = sfb.configurationManager
                config.setEndUserAcceptedVideoLicense()
                if(didJoinMeeting()){
                    self.performSegue(withIdentifier: "joinOnlineAudioVideoChat", sender: nil)
                }
            }
        }
    }
    
}
